src/CakeSlayer/Jwt.hs

```diff
@@ -0,0 +1,180 @@
+
+{- | Encoded JSON web token.
+-}
+newtype JwtToken = JwtToken
+    { unJwtToken :: Text
+    } deriving stock (Show, Generic)
+      deriving newtype (Eq, Ord, Hashable, FromField, ToField, FromHttpApiData, FromJSON, ToJSON)
+      deriving anyclass Elm
+
+{- | Stores arbitrary payload. If you want to store your custom payload, you
+need to specify two functions:
+
+1. How to encode 'JwtPayload' as 'Jwt.ClaimsMap'.
+2. How to decode 'JwtPayload from 'Jwt.ClaimsMap'.
+
+See examples in this module: 'encodeIntIdPayload', 'decodeIntIdPayload',
+'encodeTextIdPayload', 'decodeTextIdPayload'.
+-}
+newtype JwtPayload a = JwtPayload
+    { unJwtPayload :: a
+    } deriving stock (Show, Functor)
+      deriving newtype (Eq)
+
+{- | Encodes 'JwtPayload' that stores 'Int' as payload with name @id@. Use it if
+you store ids as integer values. Dual to 'decodeIntIdPayload'.
+-}
+encodeIntIdPayload :: JwtPayload Int -> Jwt.ClaimsMap
+encodeIntIdPayload = payloadToMap . Json.Number . fromIntegral . unJwtPayload
+{-# INLINE encodeIntIdPayload #-}
+
+{- | Decodes 'JwtPayload' from 'Jwt.ClaimsMap' that stores 'Int' under name
+@id@. Dual to 'encodeIntIdPayload'.
+-}
+decodeIntIdPayload :: Jwt.ClaimsMap -> Maybe (JwtPayload Int)
+decodeIntIdPayload = fmap JwtPayload . payloadFromMap (\case
+    Json.Number jwtId -> toBoundedInteger jwtId
+    _ -> Nothing)
+{-# INLINE decodeIntIdPayload #-}
+
+{- | Encodes 'JwtPayload' that stores 'Text' as payload with name @id@. Use it if
+you store ids as text or UUID values. Dual to 'decodeTextIdPayload'.
+-}
+encodeTextIdPayload :: JwtPayload Text -> Jwt.ClaimsMap
+encodeTextIdPayload = payloadToMap . Json.String . unJwtPayload
+{-# INLINE encodeTextIdPayload #-}
+
+{- | Decodes 'JwtPayload' from 'Jwt.ClaimsMap' that stores 'Int' under name
+@id@. Dual to 'encodeIntIdPayload'.
+-}
+decodeTextIdPayload :: Jwt.ClaimsMap -> Maybe (JwtPayload Text)
+decodeTextIdPayload = fmap JwtPayload . payloadFromMap (\case
+    Json.String jwtId -> Just jwtId
+    _ -> Nothing)
+{-# INLINE decodeTextIdPayload #-}
+
+-- | Creates 'Jwt.ClaimsMap' from 'Json.Value' under name @id@.
+payloadToMap :: Json.Value -> Jwt.ClaimsMap
+payloadToMap val = Jwt.ClaimsMap $ Map.fromList [("id", val)]
+{-# INLINE payloadToMap #-}
+
+{- | Decodes payload from 'Jwt.ClaimsMap' under @id@ name given 'Json.Value'
+extractor.
+-}
+payloadFromMap :: (Value -> Maybe a) -> Jwt.ClaimsMap -> Maybe a
+payloadFromMap fromValue (Jwt.ClaimsMap claimsMap) =
+    Map.lookup "id" claimsMap >>= fromValue
+{-# INLINE payloadFromMap #-}
+
+----------------------------------------------------------------------------
+-- Secret
+----------------------------------------------------------------------------
+
+{- | This monad represents effect to create and verify JWT.
+
+TODO: parametrize 'JwtPayload' when we figure out how to do this
+-}
+class Monad m => MonadJwt m where
+    mkJwtToken
+        :: Int              -- ^ Token expiry in seconds
```

Comment:
Should we add a TODO here to convert this to the `Seconds` newtype once it is in?

Reply:
I've created an issue about seconds:

* https://github.com/Holmusk/cake-slayer/issues/20